File: 5/validParentheses.py
from pdb import set_trace as bp

def isValid(line):
    stack = []
    for s in line:
        #bp()
        if (s == '{' ):
            stack.append('}')
        elif (s == '[' ):
            stack.append(']')
        elif (s == '('):
            stack.append(')')
        elif (s == '}' or s == ']' or s == ')'):
            if (len(stack) == 0 or stack.pop() != s):
                return False
    return len(stack) == 0

print(isValid('()'))
print(isValid('((})[]{})'))
print(isValid('(()[]{})'))
print(isValid('()'))

File: tsPractice/validParentheses.ts
function isValid(s: string): boolean {
    if (s.length == 1) {
        return false;
    }
    var stack = [];
    stack.push(s.charAt(0));

    for (var i = 1; i < s.length; ++i) {
        if (s.charAt(i) === '(' || s.charAt(i) === '[' || s.charAt(i) === '{') {
            stack.push(s.charAt(i))

        } else if (s.charAt(i) == ')' && stack.pop() != '(') {
            return false;
            // if (stack.pop() != '(') {
            //     return false;
            // }

        } else if (s.charAt(i) == '}' && stack.pop() != '{') {
            return false;
        } else if (s.charAt(i) == ']' && stack.pop() != '[') {
            return false;
        }

    }
    if (stack.length == 0) {
        return true;
    }
    return false;
}

console.log(isValid("(){}[]"));
console.log(isValid("()"));
console.log(isValid("(]"));

console.log(isValid("()}[]"));
console.log(isValid("()[]}"));
console.log(isValid("(["));
